File: android/app/src/main/java/com/akamrnagar/mindful/helpers/NetworkUsageHelper.java
package com.akamrnagar.mindful.helpers;

import static com.akamrnagar.mindful.utils.Extensions.getOrDefault;

import android.app.usage.NetworkStats;
import android.app.usage.NetworkStatsManager;
import android.net.ConnectivityManager;
import android.os.RemoteException;
import android.util.Log;

import androidx.annotation.NonNull;

import java.util.HashMap;


/**
 * NetworkUsageHelper is a utility class responsible for gathering network usage statistics for
 * Android applications.
 * It fetches stats about the mobile and Wi-Fi data usage of applications and organizes this
 * information into HashMap of app's uid and the usage in bytes.
 */
public class NetworkUsageHelper {
    private static final String TAG = "Mindful.NetworkUsageHelper";

    /**
     * Fetches Wi-Fi usage statistics for a specified time interval.
     *
     * @param networkStatsManager The NetworkStatsManager used to query network usage.
     * @param start               The start time of the interval in milliseconds.
     * @param end                 The end time of the interval in milliseconds.
     * @return A map with app UIDs as keys and their corresponding Wi-Fi usage in bytes as values.
     */
    @NonNull
    public static HashMap<Integer, Long> fetchWifiUsageForInterval(@NonNull NetworkStatsManager networkStatsManager, long start, long end) {
        HashMap<Integer, Long> wifiUsageMap = new HashMap<>();

        try {
            /// fetch wifi usage
            NetworkStats networkStatsWifi = networkStatsManager.querySummary(ConnectivityManager.TYPE_WIFI, null, start, end);
            NetworkStats.Bucket bucketWifi = new NetworkStats.Bucket();

            do {
                networkStatsWifi.getNextBucket(bucketWifi);
                if (wifiUsageMap.containsKey(bucketWifi.getUid())) {
                    Long previous = getOrDefault(wifiUsageMap, bucketWifi.getUid(), 0L);
                    previous += bucketWifi.getRxBytes();
                    previous += bucketWifi.getTxBytes();
                    wifiUsageMap.put(bucketWifi.getUid(), previous);
                } else {
                    long current = 0;
                    current += bucketWifi.getRxBytes();
                    current += bucketWifi.getTxBytes();
                    wifiUsageMap.put(bucketWifi.getUid(), current);
                }

            } while (networkStatsWifi.hasNextBucket());

            networkStatsWifi.close();

        } catch (RemoteException e) {
            Log.e(TAG, "fetchWifiUsageForInterval: Error in fetching wifi usage for device apps ", e);
        }

        return wifiUsageMap;
    }

    /**
     * Fetches mobile data usage statistics for a specified time interval.
     *
     * @param networkStatsManager The NetworkStatsManager used to query network usage.
     * @param start               The start time of the interval in milliseconds.
     * @param end                 The end time of the interval in milliseconds.
     * @return A map with app UIDs as keys and their corresponding mobile data usage in bytes as values.
     */
    @NonNull
    public static HashMap<Integer, Long> fetchMobileUsageForInterval(@NonNull NetworkStatsManager networkStatsManager, long start, long end) {
        HashMap<Integer, Long> mobileUsageMap = new HashMap<>();

        try {
            /// fetch mobile usage
            NetworkStats networkStatsMobile = networkStatsManager.querySummary(ConnectivityManager.TYPE_MOBILE, null, start, end);
            NetworkStats.Bucket bucketMobile = new NetworkStats.Bucket();

            do {
                networkStatsMobile.getNextBucket(bucketMobile);
                if (mobileUsageMap.containsKey(bucketMobile.getUid())) {
                    Long previous = getOrDefault(mobileUsageMap, bucketMobile.getUid(), 0L);
                    previous += bucketMobile.getRxBytes();
                    previous += bucketMobile.getTxBytes();
                    mobileUsageMap.put(bucketMobile.getUid(), previous);
                } else {
                    long current = 0;
                    current += bucketMobile.getRxBytes();
                    current += bucketMobile.getTxBytes();
                    mobileUsageMap.put(bucketMobile.getUid(), current);
                }

            } while (networkStatsMobile.hasNextBucket());

            networkStatsMobile.close();

        } catch (RemoteException e) {
            Log.e(TAG, "fetchMobileUsageForInterval: Error in fetching mobile usage for device apps ", e);
        }

        return mobileUsageMap;
    }
}

File: android/app/src/main/java/com/mindful/android/helpers/usages/NetworkUsageHelper.kt
/*
 *
 *  *
 *  *  * Copyright (c) 2024 Mindful (https://github.com/akaMrNagar/Mindful)
 *  *  * Author : Pawan Nagar (https://github.com/akaMrNagar)
 *  *  *
 *  *  * This source code is licensed under the GPL-2.0 license license found in the
 *  *  * LICENSE file in the root directory of this source tree.
 *  *
 *
 */
package com.mindful.android.helpers.usages

import android.app.usage.NetworkStats
import android.app.usage.NetworkStatsManager
import android.net.ConnectivityManager
import android.util.Log

/**
 * NetworkUsageHelper is a utility class responsible for gathering network usage statistics for
 * Android applications. It provides methods to fetch network usage data for mobile and Wi-Fi
 * connections over specified time intervals.
 */
object NetworkUsageHelper {
    private const val TAG = "Mindful.NetworkUsageHelper"

    /**
     * Fetches Wi-Fi usage statistics for a specified time interval.
     *
     * @param networkStatsManager The NetworkStatsManager used to query network usage.
     * @param start               The start time of the interval in milliseconds.
     * @param end                 The end time of the interval in milliseconds.
     * @return A map where keys are app UIDs and values are the corresponding Wi-Fi usage in KBs.
     */
    fun fetchWifiUsageForInterval(
        networkStatsManager: NetworkStatsManager,
        start: Long,
        end: Long,
    ): Map<Int, Long> =
        fetchNetworkUsageForInterval(networkStatsManager, ConnectivityManager.TYPE_WIFI, start, end)

    /**
     * Fetches mobile data usage statistics for a specified time interval.
     *
     * @param networkStatsManager The NetworkStatsManager used to query network usage.
     * @param start               The start time of the interval in milliseconds.
     * @param end                 The end time of the interval in milliseconds.
     * @return A map where keys are app UIDs and values are the corresponding mobile data usage in KBs.
     */
    fun fetchMobileUsageForInterval(
        networkStatsManager: NetworkStatsManager,
        start: Long,
        end: Long,
    ): Map<Int, Long> =
        fetchNetworkUsageForInterval(
            networkStatsManager,
            ConnectivityManager.TYPE_MOBILE,
            start,
            end
        )

    /**
     * Fetches network usage statistics for a specified network type over a given time interval.
     *
     * @param networkStatsManager The NetworkStatsManager used to query network usage.
     * @param networkType The type of network (e.g., ConnectivityManager.TYPE_WIFI or TYPE_MOBILE).
     * @param start The start time of the interval in milliseconds.
     * @param end The end time of the interval in milliseconds.
     * @return A map where keys are app UIDs and values are the corresponding data usage in KBs.
     */
    private fun fetchNetworkUsageForInterval(
        networkStatsManager: NetworkStatsManager,
        networkType: Int,
        start: Long,
        end: Long,
    ): Map<Int, Long> {
        val usageMap = mutableMapOf<Int, Long>()
        val networkStats = networkStatsManager.querySummary(networkType, null, start, end)

        try {
            val bucket = NetworkStats.Bucket()

            while (networkStats.hasNextBucket()) {
                networkStats.getNextBucket(bucket)
                val uid = bucket.uid
                usageMap[uid] = usageMap.getOrDefault(uid, 0L) + (bucket.rxBytes + bucket.txBytes)
            }

        } catch (e: Exception) {
            Log.e(
                TAG,
                "fetchNetworkUsageForInterval: Error fetching network usage for type $networkType",
                e
            )
        } finally {
            networkStats.close()
        }

        return usageMap
            .mapValues { it.value / 1024 }  // Convert bytes to KBs
            .filterValues { it > 0L }       // Remove entries with no usage
    }


}
